beacon-chain/types/active_state.go

```diff
@@ -0,0 +1,214 @@
+		blockhashes = append(blockhashes, common.BytesToHash(hash))
+	}
+	return blockhashes
+}
+
+// IsVoteCacheEmpty returns false if vote cache of an input block hash doesn't exist.
+func (a *ActiveState) isVoteCacheEmpty(blockHash [32]byte) bool {
+	_, ok := a.blockVoteCache[blockHash]
+	return ok
+}
+
+// GetBlockVoteCache returns the entire set of block vote cache.
+func (a *ActiveState) GetBlockVoteCache() map[[32]byte]*VoteCache {
+	return a.blockVoteCache
+}
+
+func (a *ActiveState) calculateNewAttestations(new []*pb.AttestationRecord, lastStateRecalc uint64) []*pb.AttestationRecord {
+	existing := a.data.PendingAttestations
+	update := make([]*pb.AttestationRecord, 0)
+	for i := 0; i < len(existing); i++ {
+		if existing[i].GetSlot() >= lastStateRecalc {
+			update = append(update, existing[i])
+		}
+	}
+	for i := 0; i < len(new); i++ {
+		update = append(update, new[i])
+	}
+
+	return update
+}
+
+func (a *ActiveState) calculateNewBlockHashes(block *Block, parentSlot uint64) ([][]byte, error) {
+	hash, err := block.Hash()
+	if err != nil {
+		return nil, err
+	}
+
+	dist := block.SlotNumber() - parentSlot
+	existing := a.data.RecentBlockHashes
+	update := existing[dist:]
+	for len(update) < 2*params.CycleLength {
+		update = append(update, hash[:])
+	}
+
+	return update, nil
+}
+
+// calculateBlockVoteCache calculates and updates active state's block vote cache.
+func (a *ActiveState) calculateNewVoteCache(block *Block, cState *CrystallizedState) (map[[32]byte]*VoteCache, error) {
+	new := voteCacheDeepCopy(a.GetBlockVoteCache())
+
+	for i := 0; i < len(block.Attestations()); i++ {
+		attestation := block.Attestations()[i]
+
+		parentHashes := a.getSignedParentHashes(block, attestation)
+		attesterIndices, err := cState.getAttesterIndices(attestation)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, h := range parentHashes {
+			// Skip calculating for this hash if the hash is part of oblique parent hashes.
+			var skip bool
+			for _, oblique := range attestation.ObliqueParentHashes {
+				if bytes.Equal(h[:], oblique) {
+					skip = true
+				}
+			}
+			if skip {
+				continue
+			}
+
+			// Initialize vote cache of a given block hash if it doesn't exist already.
+			if !a.isVoteCacheEmpty(h) {
+				new[h] = newVoteCache()
+			}
+
+			// Loop through attester indices, if the attester has voted but was not accounted for
+			// in the cache, then we add attester's index and balance to the block cache.
+			for i, attesterIndex := range attesterIndices {
+				var attesterExists bool
+				if !utils.CheckBit(attestation.AttesterBitfield, i) {
+					continue
+				}
+				for _, indexInCache := range new[h].VoterIndices {
+					if attesterIndex == indexInCache {
+						attesterExists = true
+					}
+				}
+				if !attesterExists {
+					new[h].VoterIndices = append(new[h].VoterIndices, attesterIndex)
+					new[h].VoteTotalDeposit += cState.Validators()[attesterIndex].Balance
+				}
+			}
+		}
+	}
+
+	return new, nil
```

Comment:
`new` is a go keyword - recommend to change something that clarifies exactly what it is instead